From 49c8b88671a976c788d2f0554488d96f2f91c3c2 Mon Sep 17 00:00:00 2001
From: Ravindra Lokhande <rlokhande@nvidia.com>
Date: Thu, 9 Aug 2012 22:42:05 +0530
Subject: [PATCH 1/1] tinyalsa: return card number from its name substring

tinyalsa doesn't provide any way to get card number from its name.
Added function to return card number from card name substring. This
is needed as usb audio device can be plug-in and plug-out anytime
which changes card number.

Bug 1009921
Bug 1026047

Change-Id: I2967ede50392e22f2d6f94d0b785209e45cec94f
Reviewed-on: http://git-master/r/163879

(cherry picked from commit 51c575c9e63473f1369f59f3b18d31d3a4abc3dd)

Change-Id: I2967ede50392e22f2d6f94d0b785209e45cec94f
Reviewed-on: http://git-master/r/326330
Tested-by: Jean Huang <jeanh@nvidia.com>
Reviewed-by: Peter Zu <pzu@nvidia.com>
Reviewed-by: Jean Huang <jeanh@nvidia.com>
---
 include/tinyalsa/asoundlib.h |    3 +++
 mixer.c                      |   32 ++++++++++++++++++++++++++++++++
 2 files changed, 35 insertions(+), 0 deletions(-)

diff --git a/include/tinyalsa/asoundlib.h b/include/tinyalsa/asoundlib.h
index ec526e6..48a181a 100644
--- a/include/tinyalsa/asoundlib.h
+++ b/include/tinyalsa/asoundlib.h
@@ -259,6 +259,9 @@ int mixer_ctl_set_enum_by_string(struct mixer_ctl *ctl, const char *string);
 int mixer_ctl_get_range_min(struct mixer_ctl *ctl);
 int mixer_ctl_get_range_max(struct mixer_ctl *ctl);
 
+/* Get sound card number using sub string */
+int snd_card_get_card_id_from_sub_string(const char *card_name);
+
 #if defined(__cplusplus)
 }  /* extern "C" */
 #endif
diff --git a/mixer.c b/mixer.c
index 4568cca..54d826e 100644
--- a/mixer.c
+++ b/mixer.c
@@ -495,3 +495,35 @@ int mixer_ctl_set_enum_by_string(struct mixer_ctl *ctl, const char *string)
     return -EINVAL;
 }
 
+
+int snd_card_get_card_id_from_sub_string(const char *card_name)
+{
+    int card;
+    int fd;
+    char fn[256];
+    struct snd_ctl_card_info info;
+
+    if (card_name == NULL)
+        return -EINVAL;
+
+    for (card = 0; card < 8; card++) {
+        snprintf(fn, sizeof(fn), "/dev/snd/controlC%u", card);
+        fd = open(fn, O_RDONLY);
+        if (fd < 0)
+            continue;
+        if (ioctl(fd, SNDRV_CTL_IOCTL_CARD_INFO, &info) < 0) {
+            close(fd);
+            continue;
+        }
+        close(fd);
+
+        /* search by card id */
+        if (strcasestr((const char *)info.id, card_name))
+            return info.card;
+
+        /* search by card name */
+        if (strcasestr((const char *)info.name, card_name))
+            return info.card;
+    }
+    return -EINVAL;
+}
-- 
1.6.4.2

